File: ios/TodayFlow/TodayFlow/Services/TodayGuideActionable.swift
import Foundation

/// Паритет с `frontend/src/components/today/todayGuideActionable.ts` — поля guide narrative после ритуала.
struct SphereTriadGuideRow: Equatable {
    let area: String
    let stance: String
    let line: String
}

enum TodayGuideActionable {
    /// O2: паритет бэка `narrative_hierarchy` в guide payload (`primary_anchor` = `day_engine_brief`).
    struct NarrativeHierarchyDisplay: Equatable {
        let contractVersion: String
        let primaryAnchorKey: String
    }

    static func narrativeHierarchyDisplay(from payload: [String: JSONValue]?) -> NarrativeHierarchyDisplay? {
        guard let payload, case let .object(h) = payload["narrative_hierarchy"] ?? .null else { return nil }
        guard h["contract_version"]?.stringValue == "narrative_hierarchy_v0" else { return nil }
        guard h["primary_anchor"]?.stringValue == "day_engine_brief" else { return nil }
        return NarrativeHierarchyDisplay(contractVersion: "narrative_hierarchy_v0", primaryAnchorKey: "day_engine_brief")
    }

    /// DE-13 v5: паритет `guide_pipeline` / `guide_contract_v2`.
    struct GuidePipelineDisplay: Equatable {
        let contractVersion: String
        let generationMode: String
        let coreSource: String?
    }

    static func guidePipelineDisplay(from payload: [String: JSONValue]?) -> GuidePipelineDisplay? {
        guard let payload, payload["contract_version"]?.stringValue == "guide_contract_v2" else { return nil }
        guard case let .object(p) = payload["guide_pipeline"] ?? .null else { return nil }
        guard p["contract_version"]?.stringValue == "guide_pipeline_v0" else { return nil }
        let mode = p["generation_mode"]?.stringValue ?? ""
        guard mode == "funnel" || mode == "monolith" else { return nil }
        var coreSource: String?
        if case let .object(steps) = p["steps"] ?? .null,
           case let .object(core) = steps["core_text"] ?? .null {
            coreSource = core["source"]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines)
            if coreSource?.isEmpty == true { coreSource = nil }
        }
        return GuidePipelineDisplay(contractVersion: "guide_pipeline_v0", generationMode: mode, coreSource: coreSource)
    }

    /// Паритет с веб `guidePayload.day_engine_brief`: якорь и подсказки для блока «Опора дня».
    struct DayEngineBriefDisplay: Equatable {
        let anchor: String
        let hints: [String]
    }

    static func dayEngineBriefDisplay(from payload: [String: JSONValue]?) -> DayEngineBriefDisplay? {
        guard let payload, case let .object(b) = payload["day_engine_brief"] ?? .null else { return nil }
        let anchor = (b["anchor_summary"]?.stringValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !anchor.isEmpty else { return nil }
        var hints: [String] = []
        for key in ["do_hint", "avoid_hint", "tempo_hint"] {
            if let s = b[key]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines), !s.isEmpty {
                hints.append(s)
            }
        }
        return DayEngineBriefDisplay(anchor: anchor, hints: hints)
    }

    /// Паритет с веб `parseDayModelBriefFromGuide` — `day_model` (`day_model_v0`), поле `strategy.one_focus`.
    struct DayModelBriefDisplay: Equatable {
        let oneFocus: String
        let vectorSummary: String?
    }

    static func dayModelBriefDisplay(from payload: [String: JSONValue]?) -> DayModelBriefDisplay? {
        guard let payload, case let .object(dm) = payload["day_model"] ?? .null else { return nil }
        guard dm["contract_version"]?.stringValue == "day_model_v0" else { return nil }
        let vectorSummary: String?
        if case let .object(v) = dm["vector"] ?? .null {
            let s = (v["summary"]?.stringValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            vectorSummary = s.isEmpty ? nil : s
        } else {
            vectorSummary = nil
        }
        let oneFocus: String
        if case let .object(st) = dm["strategy"] ?? .null {
            let s = (st["one_focus"]?.stringValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            guard !s.isEmpty else { return nil }
            oneFocus = s
        } else {
            return nil
        }
        return DayModelBriefDisplay(oneFocus: oneFocus, vectorSummary: vectorSummary)
    }

    /// Паритет с веб `parseCoreMessageForUi`: строка или объект `{ body, risk?, best_move?, headline? }`.
    static func coreMessageDisplayString(from payload: [String: JSONValue]?) -> String {
        guard let raw = payload?["core_message"] else { return "" }
        if case let .string(s) = raw {
            return s.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if case let .object(o) = raw {
            let body = (o["body"]?.stringValue ?? o["main_text"]?.stringValue ?? o["message"]?.stringValue ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            guard !body.isEmpty else { return "" }
            var headline = (o["headline"]?.stringValue ?? o["title"]?.stringValue ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if headline.isEmpty { headline = "" }
            var risk = (o["risk"]?.stringValue ?? o["main_risk"]?.stringValue ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            var bestMove = (o["best_move"]?.stringValue ?? o["first_move"]?.stringValue ?? o["action_hint"]?.stringValue ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !headline.isEmpty, TodayPersonalDayRhythm.textsSemanticallyRedundant(headline, body) { headline = "" }
            if !risk.isEmpty, TodayPersonalDayRhythm.textsSemanticallyRedundant(risk, body) { risk = "" }
            if !bestMove.isEmpty, TodayPersonalDayRhythm.textsSemanticallyRedundant(bestMove, body) { bestMove = "" }
            if !risk.isEmpty, !bestMove.isEmpty, TodayPersonalDayRhythm.textsSemanticallyRedundant(risk, bestMove) { risk = "" }
            if !bestMove.isEmpty, !headline.isEmpty, TodayPersonalDayRhythm.textsSemanticallyRedundant(bestMove, headline) { bestMove = "" }
            var parts: [String] = []
            if !headline.isEmpty { parts.append(headline) }
            parts.append(body)
            if !risk.isEmpty { parts.append("\(TodayRitualCopy.heroRiskLabel): \(risk)") }
            if !bestMove.isEmpty { parts.append("\(TodayRitualCopy.heroBestMoveLabel): \(bestMove)") }
            return parts.joined(separator: "\n\n")
        }
        return ""
    }

    static func coreMessageParagraphs(from payload: [String: JSONValue]?) -> [String] {
        let flat = coreMessageDisplayString(from: payload)
        if flat.isEmpty { return [] }
        return flat
            .components(separatedBy: "\n\n")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .prefix(4)
            .map { String($0) }
    }

    static func parseActionOptions(from payload: [String: JSONValue]?) -> [String] {
        guard let payload, case let .array(items) = payload["action_options"] ?? .null else { return [] }
        var out: [String] = []
        for item in items.prefix(3) {
            if let s = item.stringValue, !s.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                out.append(s.trimmingCharacters(in: .whitespacesAndNewlines))
            } else if case let .object(o) = item {
                let title = (o["title"]?.stringValue ?? o["label"]?.stringValue ?? o["text"]?.stringValue ?? "")
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                if !title.isEmpty { out.append(title) }
            }
        }
        return out
    }

    /// O9: один канонический «главный шаг» — `best_move` → первый `action_options` → первый `do_items` (паритет веб `guideCanonicalPrimaryStepLine`).
    static func guideCanonicalPrimaryStepLine(payload: [String: JSONValue]?, doItems: [String], fallback: String) -> String {
        if let payload, case let .object(o) = payload["core_message"] ?? .null {
            let body = (o["body"]?.stringValue ?? o["main_text"]?.stringValue ?? o["message"]?.stringValue ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            var bm = (o["best_move"]?.stringValue ?? o["first_move"]?.stringValue ?? o["action_hint"]?.stringValue ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !bm.isEmpty, !body.isEmpty, TodayPersonalDayRhythm.textsSemanticallyRedundant(bm, body) { bm = "" }
            if !bm.isEmpty { return bm }
        }
        let opts = parseActionOptions(from: payload)
        if let f = opts.first { return f }
        if let d = doItems.first(where: { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }) { return d }
        let t = fallback.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? TodayWebGuideSectionCopy.guidePrimaryDoFallback : t
    }

    static func parseSupportHooks(from payload: [String: JSONValue]?) -> [String] {
        guard let payload, case let .array(items) = payload["support_hooks"] ?? .null else { return [] }
        return items.compactMap { $0.stringValue }.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }.prefix(3).map { String($0) }
    }

    static func parseSphereTriad(from payload: [String: JSONValue]?) -> [SphereTriadGuideRow]? {
        guard let payload, case let .array(items) = payload["sphere_triad"] ?? .null, items.count == 3 else { return nil }
        var rows: [SphereTriadGuideRow] = []
        for item in items {
            guard case let .object(o) = item else { return nil }
            let area = (o["area"]?.stringValue ?? "").lowercased()
            let stance = (o["stance"]?.stringValue ?? "").lowercased()
            let line = o["line"]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard ["work", "love", "money"].contains(area), ["up", "down", "neutral"].contains(stance), line.count > 4 else { return nil }
            rows.append(SphereTriadGuideRow(area: area, stance: stance, line: line))
        }
        guard Set(rows.map(\.area)).count == 3 else { return nil }
        return rows
    }
}

File: frontend/src/components/today/todayGuideActionable.ts
import type { TodayFourArea } from "@/components/today/todayFourAreas";
import { RITUAL_COPY, textsSemanticallyRedundant } from "@/components/today/todayRitualCopy";

export type SphereTriadRow = {
  area: "work" | "love" | "money";
  stance: "up" | "down" | "neutral";
  line: string;
};

function narrativeString(payload: Record<string, unknown> | null | undefined, key: string): string | null {
  if (!payload || typeof payload !== "object") return null;
  const v = payload[key];
  return typeof v === "string" && v.trim() ? v.trim() : null;
}

export function parseCoreMessageFromGuide(payload: Record<string, unknown> | null | undefined): string | null {
  const raw = payload?.["core_message"];
  if (raw && typeof raw === "object" && !Array.isArray(raw)) {
    const o = raw as Record<string, unknown>;
    const body =
      (typeof o.body === "string" && o.body.trim()) ||
      (typeof o.main_text === "string" && o.main_text.trim()) ||
      (typeof o.message === "string" && o.message.trim()) ||
      "";
    if (body) {
      const parts: string[] = [body];
      const risk = (typeof o.risk === "string" && o.risk.trim()) || (typeof o.main_risk === "string" && o.main_risk.trim()) || "";
      const move =
        (typeof o.best_move === "string" && o.best_move.trim()) ||
        (typeof o.first_move === "string" && o.first_move.trim()) ||
        "";
      if (risk) parts.push(`Риск: ${risk}`);
      if (move) parts.push(`Лучший ход: ${move}`);
      return parts.join("\n\n");
    }
  }
  const flat = narrativeString(payload, "core_message");
  return flat || null;
}

/** Объект `core_message` из API (предпочтительный контракт) или разбор строки. */
export type ParsedCoreMessageForUi =
  | {
      kind: "structured";
      headline?: string;
      body: string;
      risk?: string;
      best_move?: string;
    }
  | { kind: "paragraphs"; paragraphs: string[] };

export function parseCoreMessageForUi(payload: Record<string, unknown> | null | undefined): ParsedCoreMessageForUi | null {
  const raw = payload?.["core_message"];
  if (raw && typeof raw === "object" && !Array.isArray(raw)) {
    const o = raw as Record<string, unknown>;
    const body =
      (typeof o.body === "string" && o.body.trim()) ||
      (typeof o.main_text === "string" && o.main_text.trim()) ||
      (typeof o.message === "string" && o.message.trim()) ||
      "";
    if (body) {
      let headline =
        (typeof o.headline === "string" && o.headline.trim()) || (typeof o.title === "string" && o.title.trim()) || undefined;
      let risk =
        (typeof o.risk === "string" && o.risk.trim()) || (typeof o.main_risk === "string" && o.main_risk.trim()) || undefined;
      let best_move =
        (typeof o.best_move === "string" && o.best_move.trim()) ||
        (typeof o.first_move === "string" && o.first_move.trim()) ||
        (typeof o.action_hint === "string" && o.action_hint.trim()) ||
        undefined;
      if (headline && textsSemanticallyRedundant(headline, body)) headline = undefined;
      if (risk && textsSemanticallyRedundant(risk, body)) risk = undefined;
      if (best_move && textsSemanticallyRedundant(best_move, body)) best_move = undefined;
      if (risk && best_move && textsSemanticallyRedundant(risk, best_move)) risk = undefined;
      if (best_move && headline && textsSemanticallyRedundant(best_move, headline)) best_move = undefined;
      return { kind: "structured", headline, body, risk, best_move };
    }
  }
  const flat = narrativeString(payload, "core_message");
  if (!flat) return null;
  const paragraphs = flat
    .split(/\n\s*\n/)
    .map((p) => p.trim())
    .filter(Boolean)
    .slice(0, 4);
  return paragraphs.length ? { kind: "paragraphs", paragraphs } : null;
}

export type TodayGuideActionOption = {
  title: string;
  reason?: string;
  estimated_minutes?: number;
  entity_kind?: string;
};

function narrativeStringFromUnknown(v: unknown): string | null {
  return typeof v === "string" && v.trim() ? v.trim() : null;
}

export function parseActionOptionsRich(payload: Record<string, unknown> | null | undefined): TodayGuideActionOption[] {
  if (!payload || typeof payload !== "object") return [];
  const raw = payload["action_options"];
  if (!Array.isArray(raw)) return [];
  const out: TodayGuideActionOption[] = [];
  for (const x of raw) {
    if (typeof x === "string") {
      const t = x.trim();
      if (t) out.push({ title: t });
    } else if (x && typeof x === "object" && !Array.isArray(x)) {
      const o = x as Record<string, unknown>;
      const title =
        narrativeStringFromUnknown(o.title) ||
        narrativeStringFromUnknown(o.label) ||
        narrativeStringFromUnknown(o.text);
      if (title) {
        out.push({
          title,
          reason: narrativeStringFromUnknown(o.reason) || narrativeStringFromUnknown(o.why) || undefined,
          estimated_minutes:
            typeof o.estimated_minutes === "number" && Number.isFinite(o.estimated_minutes)
              ? o.estimated_minutes
              : typeof o.minutes === "number" && Number.isFinite(o.minutes)
                ? o.minutes
                : undefined,
          entity_kind:
            typeof o.entity_kind === "string"
              ? o.entity_kind
              : typeof o.creates === "string"
                ? o.creates
                : undefined,
        });
      }
    }
    if (out.length >= 3) break;
  }
  return out;
}

export function parseActionOptionsFromGuide(payload: Record<string, unknown> | null | undefined): string[] {
  return parseActionOptionsRich(payload).map((o) => o.title);
}

function doItemsFromGuidePayload(payload: Record<string, unknown> | null | undefined): string[] {
  if (!payload || typeof payload !== "object") return [];
  const raw = payload["do_items"];
  if (!Array.isArray(raw)) return [];
  return raw.map((x) => (typeof x === "string" ? x.trim() : "")).filter(Boolean).slice(0, 6);
}

/**
 * O9: один канонический «главный шаг» для UI — `best_move` → первый `action_options` → первый `do_items`.
 * Паритет iOS `TodayGuideActionable.guideCanonicalPrimaryStepLine`.
 */
export function guideCanonicalPrimaryStepLine(
  payload: Record<string, unknown> | null | undefined,
  doItemsFallback: readonly string[],
  literalFallback: string,
): string {
  const core = parseCoreMessageForUi(payload);
  if (core?.kind === "structured") {
    const bm = core.best_move?.trim();
    if (bm) return bm;
  }
  const opts = parseActionOptionsRich(payload);
  if (opts.length > 0 && opts[0].title.trim()) return opts[0].title.trim();
  const dos = doItemsFromGuidePayload(payload);
  if (dos.length > 0) return dos[0].trim();
  for (const x of doItemsFallback) {
    const t = String(x || "").trim();
    if (t) return t;
  }
  return literalFallback.trim() || RITUAL_COPY.guidePrimaryDoFallback;
}

export function parseSupportHooksFromGuide(payload: Record<string, unknown> | null | undefined): string[] {
  if (!payload || typeof payload !== "object") return [];
  const raw = payload["support_hooks"];
  if (!Array.isArray(raw)) return [];
  return raw.map((x) => (typeof x === "string" ? x.trim() : "")).filter(Boolean).slice(0, 3);
}

/** O2: паритет бэка `narrative_hierarchy` в guide payload. */
export type NarrativeHierarchyForUi = {
  contractVersion: "narrative_hierarchy_v0";
  primaryAnchorKey: "day_engine_brief";
};

export function parseNarrativeHierarchyFromGuide(
  payload: Record<string, unknown> | null | undefined,
): NarrativeHierarchyForUi | null {
  if (!payload || typeof payload !== "object") return null;
  const raw = payload["narrative_hierarchy"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const h = raw as Record<string, unknown>;
  if (h.contract_version !== "narrative_hierarchy_v0") return null;
  if (h.primary_anchor !== "day_engine_brief") return null;
  return { contractVersion: "narrative_hierarchy_v0", primaryAnchorKey: "day_engine_brief" };
}

/** DE-13 v5: паритет бэка `guide_pipeline` в guide payload (`guide_contract_v2`). */
export type GuidePipelineForUi = {
  contractVersion: "guide_pipeline_v0";
  generationMode: "funnel" | "monolith";
  coreSource?: string;
};

export function parseGuidePipelineFromGuide(
  payload: Record<string, unknown> | null | undefined,
): GuidePipelineForUi | null {
  if (!payload || typeof payload !== "object") return null;
  if (payload.contract_version !== "guide_contract_v2") return null;
  const raw = payload["guide_pipeline"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const p = raw as Record<string, unknown>;
  if (p.contract_version !== "guide_pipeline_v0") return null;
  const mode = p.generation_mode;
  if (mode !== "funnel" && mode !== "monolith") return null;
  const steps = p.steps;
  let coreSource: string | undefined;
  if (steps && typeof steps === "object" && !Array.isArray(steps)) {
    const core = (steps as Record<string, unknown>).core_text;
    if (core && typeof core === "object" && !Array.isArray(core)) {
      const src = (core as Record<string, unknown>).source;
      if (typeof src === "string" && src.trim()) coreSource = src.trim();
    }
  }
  return { contractVersion: "guide_pipeline_v0", generationMode: mode, coreSource };
}

/** Паритет с `day_engine_brief` в payload guide. */
export type DayEngineBriefForUi = {
  anchor: string;
  hints: string[];
};

export function parseDayEngineBriefFromGuide(payload: Record<string, unknown> | null | undefined): DayEngineBriefForUi | null {
  if (!payload || typeof payload !== "object") return null;
  const raw = payload["day_engine_brief"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const b = raw as Record<string, unknown>;
  const anchor = typeof b.anchor_summary === "string" ? b.anchor_summary.trim() : "";
  if (!anchor) return null;
  const hints = (["do_hint", "avoid_hint", "tempo_hint"] as const)
    .map((k) => (typeof b[k] === "string" ? (b[k] as string).trim() : ""))
    .filter(Boolean);
  return { anchor, hints };
}

/** Паритет с `day_model_v0` в payload guide / DayContext.layers.day_model. */
export type DayModelBriefForUi = {
  contractVersion: string;
  oneFocus: string;
  vectorSummary?: string;
  tensionSummary?: string;
  riskSummary?: string;
  tempoLabel?: string;
};

function nestedString(o: unknown, path: [string, string]): string | undefined {
  if (!o || typeof o !== "object" || Array.isArray(o)) return undefined;
  const inner = (o as Record<string, unknown>)[path[0]];
  if (!inner || typeof inner !== "object" || Array.isArray(inner)) return undefined;
  const v = (inner as Record<string, unknown>)[path[1]];
  return typeof v === "string" && v.trim() ? v.trim() : undefined;
}

export function parseDayModelBriefFromGuide(payload: Record<string, unknown> | null | undefined): DayModelBriefForUi | null {
  if (!payload || typeof payload !== "object") return null;
  const raw = payload["day_model"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const dm = raw as Record<string, unknown>;
  if (dm["contract_version"] !== "day_model_v0") return null;
  const oneFocus = nestedString(dm, ["strategy", "one_focus"]);
  if (!oneFocus) return null;
  const vectorSummary = nestedString(dm, ["vector", "summary"]);
  const tensionSummary = nestedString(dm, ["tension", "summary"]);
  const riskSummary = nestedString(dm, ["risk", "summary"]);
  const scales = dm["scales"];
  let tempoLabel: string | undefined;
  if (scales && typeof scales === "object" && !Array.isArray(scales)) {
    const t = (scales as Record<string, unknown>)["tempo"];
    if (typeof t === "string" && t.trim()) tempoLabel = t.trim();
  }
  return {
    contractVersion: "day_model_v0",
    oneFocus,
    vectorSummary,
    tensionSummary,
    riskSummary,
    tempoLabel,
  };
}

function isTriadRow(x: unknown): x is SphereTriadRow {
  if (!x || typeof x !== "object") return false;
  const r = x as Record<string, unknown>;
  const area = String(r.area || "").toLowerCase();
  const stance = String(r.stance || "").toLowerCase();
  const line = String(r.line || "").trim();
  return (area === "work" || area === "love" || area === "money") && (stance === "up" || stance === "down" || stance === "neutral") && line.length > 4;
}

export function parseSphereTriadFromGuide(payload: Record<string, unknown> | null | undefined): SphereTriadRow[] | null {
  if (!payload || typeof payload !== "object") return null;
  const raw = payload["sphere_triad"];
  if (!Array.isArray(raw) || raw.length !== 3) return null;
  const rows = raw.map((x) => (isTriadRow(x) ? x : null)).filter((x): x is SphereTriadRow => x != null);
  if (rows.length !== 3) return null;
  const areas = new Set(rows.map((r) => r.area));
  if (areas.size !== 3) return null;
  return rows;
}

/** Если в payload нет sphere_triad — строим из трёх сфер по score (сильная / слабая / нейтральная). */
export function sphereTriadFallbackFromAreas(areas: TodayFourArea[], relationshipLabel: string): SphereTriadRow[] {
  const pick = (id: "work" | "love" | "money") => areas.find((a) => a.id === id);
  const triple = [
    { id: "work" as const, a: pick("work") },
    { id: "love" as const, a: pick("love") },
    { id: "money" as const, a: pick("money") },
  ].filter((x) => x.a) as { id: "work" | "love" | "money"; a: TodayFourArea }[];
  if (triple.length !== 3) return [];
  const sorted = [...triple].sort((x, y) => y.a.score - x.a.score);
  const strong = sorted[0]!;
  const weak = sorted[2]!;
  const mid = sorted[1]!;
  const label = (id: "work" | "love" | "money", a: TodayFourArea) => (id === "love" ? relationshipLabel : a.label);
  const line = (id: "work" | "love" | "money", a: TodayFourArea, stance: SphereTriadRow["stance"]) => {
    const head = label(id, a);
    const tail = a.todayHeadline?.trim() || a.insight?.trim() || "";
    if (stance === "up") return tail ? `${head} — сейчас лучшее место для движения: ${tail}` : `${head} — сейчас лучшее место для движения.`;
    if (stance === "down")
      return tail ? `${head} — аккуратно, не дави: ${tail}` : `${head} — аккуратно, не дави — лучше сказать прямо, чем угадывать.`;
    return tail ? `${head} — нейтрально: ${tail}` : `${head} — нейтрально, без лишнего импульса.`;
  };
  return [
    { area: strong.id, stance: "up", line: line(strong.id, strong.a, "up") },
    { area: weak.id, stance: "down", line: line(weak.id, weak.a, "down") },
    { area: mid.id, stance: "neutral", line: line(mid.id, mid.a, "neutral") },
  ];
}
